### ISSUE TITLE
Does correcting the type head's own imbalance turn flat slots into planes without costing the surplus?

### ISSUE THREAD

**Issue by danvisai** [danvisai/SDFusion#138]
Part of #1.

*Follow-up to [#132](https://github.com/danvisai/SDFusion/issues/132), which changed the assignment head and the pitch decode, named the **type** head as the untouched binding constraint on its own KILL, and deliberately did not touch it — "one change per head was the whole design of this arm." Fourth arm on the program formulation [#6](https://github.com/danvisai/SDFusion/issues/6) chose.*

> ⚠️ **Filed retroactively on 2026-09-02, after the run.** The work was executed and written up before this issue existed: the write-up was authored as `137-type-head-imbalance.md`, against an issue **#137 that was never created** and has since been taken by `PR #137`. The question and bar below are transcribed from the module docstring in `scripts/foundations/train_height_map_generator.py`. ⚠️ That docstring, the run's code and its artifacts are **uncommitted** on `massing-solid-gate-retrain` at filing time, so the pre-registration is **not independently verifiable in git history** — it is recorded here on the strength of the write-up, not of a timestamp. Noted so this is never later read as having been pre-registered in the tracker.

## Question

#132 left `planar_fraction` at **0.12** — 61% of the arm's used slots typed `Layer`, 59.2% compiling flat — and named that the binding constraint on the fourth clause of `PROGRAM_BAR` without asking *why* the type head reads that way.

**Is the type head diffuse, confidently wrong, or tracking a label that is itself steeply slot-index-conditional? And whichever it is, does the same fix #132 used on the assignment head — a training-side logit adjustment, not a decode-side one — work on this head too?**

## The free diagnosis first, in #132's own order

#132's most valuable move was measuring the decode **before** spending a run, and finding the post-hoc correction destroyed the dominant slot. The same three questions transfer:

* Is the label itself slot-index-conditional? Slots are canonicalised by **area** ([#6](https://github.com/danvisai/SDFusion/issues/6)), so a building's biggest region should be a pitch more often than its smallest — if so, a single pooled scalar prior would miss it and the prior must be **per slot**.
* Is the head blind, or is a plain argmax at a fixed 0.5 threshold the wrong decision rule for a low base rate?
* Does the decode-side fix work — and if it does not, does it fail the *same way* #132's did, by relabelling the building rather than reading it better?

⚠️ Whichever way this lands, the decode-side read must be refuted **before** the run, not after, or the loss-side choice is selected on the answer.

## Pre-registered bar

[#6](https://github.com/danvisai/SDFusion/issues/6)'s `PROGRAM_BAR`, unchanged for the fourth time, machine-checked by `verdict()` on the same **411 carve-needing** rows ([#126](https://github.com/danvisai/SDFusion/issues/126)):

    PASS   ops <= 3.0   AND   planar >= 0.40   AND   extra < 0.0603
    GUARD  collapse <= 1-NN's 0.1582            AND  vs_input < 0.98
    KILL   planar <= 0.20

Nothing else moves: same 40 epochs, same seed, same `plane_head class`, same `PLANE_DECODE` (pitch q0.25), same `assign_prior` at `ASSIGN_TEMPERATURE` 1.0, same selection rule (validation `missing + extra`).

⚠️ **Pre-registered prediction, so this is falsifiable:** the type fix should raise `planar_fraction` and the `Ramp`-typed share of used slots **without** moving `extra`/`missing`/collapse by much, because the assignment head and the pitch decode are untouched. If `extra` or the collapse rate move by more than a rounding amount instead, that is a result about the two heads' losses **interacting**, not about either head read in isolation.

## Judged on

* **`vs_input` and the collapse rate beside every number on every table**, diagnostics included — #129's process failure was a decode table published without them, on which a no-op ranked best.
* **`slot_usage` beside `dl_ops`** — 1.0 ops at 0.90 slots and 1.0 ops at 3.0 slots are different results.
* **A per-slot breakdown**, not a pooled scalar, for both the prior and the recall — a pooled figure cannot see a gradient across slot index.
* ⚠️ A montage before this arm is considered for anything beyond the record: [#127](https://github.com/danvisai/SDFusion/issues/127) and [#131](https://github.com/danvisai/SDFusion/issues/131) both had a picture overturn a scorecard.

**Comment by danvisai** [comment]
## Resolution

**A fourth NOT MET — and the first arm on this map's record to clear BOTH form clauses. The type head's imbalance was a calibration problem, fixable in the loss exactly as diagnosed, and the arm it produced is a worse building.**

Write-up: `docs/wayfinding/solid-first-subtractive-modeling/138-type-head-imbalance.md` on `massing-solid-gate-retrain`. ⚠️ **Uncommitted at the time of writing** — code, artifacts and write-up are all in the working tree only.

### 🔑 The free question came first, and the decode-side fix was refuted before the run

`type_prior` shows the label is steeply slot-index-conditional — slots are canonicalised by **area** ([#6](https://github.com/danvisai/SDFusion/issues/6)), so the biggest region is a pitch more often than not and the smallest is almost always a flat setback. Over 34,909 training rows, `Ramp` share by slot: **0.594 / 0.523 / 0.323 / 0.134**. A single pooled scalar prior would have missed the gradient entirely.

`type_collapse` on #132's checkpoint, **before** this fix was written:

| slot | label Ramp share | recall(Ramp) | recall(Layer) | p(Ramp \| Ramp label) | p(Ramp \| Layer label) |
|---|---|---|---|---|---|
| 0 | 0.555 | 0.741 | 0.721 | 0.691 | 0.389 |
| 1 | 0.501 | 0.626 | 0.757 | 0.611 | 0.381 |
| 2 | 0.299 | 0.357 | 0.939 | 0.421 | 0.246 |
| 3 | 0.105 | **0.087** | 0.995 | 0.235 | 0.128 |

🔑 **The head is not blind** — `p(Ramp | Ramp label)` exceeds `p(Ramp | Layer label)` at every slot. A plain argmax at a fixed 0.5 threshold is the wrong decision rule for a base rate that low: slot 3 is `Ramp` only 10.5% of the time, so a posterior averaging 0.235 on true Ramps almost never crosses 0.5 and recall craters to 8.7%. **Calibration, not representation** — the same shape #132 found on the assignment head.

⚠️ **And the decode-side fix is refuted the same way and for the same reason.** Dividing each slot's posterior by its own prior before the argmax recovers slot 3's Ramp recall 0.087 → 0.957 and pays Layer recall 0.995 → **0.413** at that slot, overall accuracy 0.7576 → 0.6757. It relabels the building's flat regions as pitched — `decode_assignment`'s failure mode arriving at the second head. Pinned by `TestTypeStats.test_the_balanced_read_can_flip_a_slot_the_argmax_loses`; the served decode stays a plain argmax.

**So the fix went into the LOSS**, where it can change what is learned: `tau * log(prior[k, c])` on slot k's class-c TYPE logit during training (`TYPE_TEMPERATURE` fixed a priori at 1.0, not swept). The prior travels with the checkpoint so a re-scored old checkpoint cannot present itself as having trained under a correction it did not have.

### Result, on the 411 carve-needing rows, judged by #6's unchanged `PROGRAM_BAR`

Selected **epoch 21 of 40** — the same epoch #132's run selected.

| arm | `missing` | `extra` | `vs_input` | collapse | ops | **planar** | *(3D IoU)* |
|---|---|---|---|---|---|---|---|
| the real building | — | — | — | — | **2.0** | **0.50** | — |
| 1-NN retrieval *(the guard)* | 0.0257 | 0.1031 | 0.8743 | **0.1582** | 2.0 | 0.17 | *0.8355* |
| CE + median *(#127, served)* | 0.0385 | **0.0603** | 0.8432 | 0.0268 | 6.0 | 0.20 | *0.8948* |
| `heightmap_program_adj` *(#132)* | 0.0659 | 0.0832 | 0.8470 | 0.2579 | 1.0 | 0.12 | *0.8080* |
| **`heightmap_program_typeadj` *(this arm)*** | 0.0835 | 0.0938 | 0.8163 | **0.2774** | 1.0 | **0.50** | *0.7960* |

    PASS   ops <= 3.0            1.0     ✔
           planar >= 0.40        0.50    ✔
           extra < 0.0603      0.0938    ✘
    GUARD  collapse <= 0.1582  0.2774    ✘
           vs_input < 0.98     0.8163    ✔
    KILL   planar <= 0.20        0.50    -> not fired

Evaluated by `verdict()`, not by this table. **Verdict: NOT MET** — but for a *different reason* than #132: #132's `killed_flat` clause fired; this arm clears the KILL by a wide margin and fails on `beats_served_extra` and the collapse `GUARD` instead.

### ✅ What worked, exactly as diagnosed

**`planar_fraction` 0.12 → 0.50 — matching the real building's own ratio exactly, and the best number ever measured on a trained arm on this map.** Every earlier program arm scored 0.00 (#6, #129), 0.12 (#132) or 0.17–0.25 (#127's per-column arms). `used_slots_typed_ramp` 0.390 → **0.570**.

`type_collapse` on this arm's own checkpoint, plain argmax: slot 3's Ramp recall **0.087 → 0.739** — most of the way to the decode-side read's 0.957 and at a far better price (Layer recall 0.643 against that read's 0.413). Slot 2 moves 0.357 → 0.595. Slots 0 and 1, already near 50/50, barely move — exactly the shape `type_prior`'s per-slot design predicts.

### 🔑🔑 Why it still fails: the same correction that recalls a real Ramp mistypes a real Layer

`compile_program` reads a slot's continuous plane parameters **only when its type says `Ramp`**, so a genuinely flat slot now mistyped `Ramp` gets a real non-zero pitch compiled into a roof that should have stayed level. Every recovered Ramp recall is paid for in Layer recall at the same slot, and slots 2 and 3 are large fractions of the population (281 and 219 of 411 buildings have an active slot there).

| label slot count | ops (#132 → this) | planar (#132 → this) | extra (#132 → this) |
|---|---|---|---|
| 1 slot | 1.0 → 1.0 | 1.00 → 1.00 | 0.0257 → 0.0309 |
| 2 slots | 1.0 → 1.0 | 1.00 → 0.60 | 0.0646 → 0.0696 |
| 3 slots | 1.0 → 1.0 | 0.33 → 0.33 | 0.0861 → 0.0807 |
| 4 slots | 1.0 → 1.0 | **0.00 → 0.40** | 0.1360 → 0.1404 |

The 4-slot bucket is where `planar` moves most, and it is both the largest bucket (219 of 411, 53%) and the one already carrying the worst collapse before this fix. **Buying planes there is buying them in the population least able to absorb a wrong one.**

⚠️ **`ops` is completely flat at 1.0 in every bucket.** The type fix changes what a slot *is* once one has been drawn; it does nothing to how many the untouched assignment head draws.

### The pre-registered prediction was half wrong, and that is the finding

Predicted: planar up, `extra`/`missing`/collapse roughly still. **`planar` moved far more than "raised" (0.12 → 0.50) and the surplus numbers did not stay put — they got worse by more than a rounding amount.** The prediction named that outcome as meaning "the two heads' losses interact," and that is what happened.

### Does not settle

* ⚠️ Whether a smaller `TYPE_TEMPERATURE` trades some planar gain back for surplus. 1.0 is the full pre-registered adjustment, deliberately not swept — sweeping after seeing this result would be selecting on the answer.
* ⚠️ **No montage** (`--montage 0`, matching #132's command). #127 and #131 both had a picture overturn a scorecard; a visual check of where the new false-`Ramp` slots land is open before this arm is considered for anything beyond the record.
* The assignment head's own diffuseness is untouched here and is very likely why the type errors land where they do — inferred from the bucket table, not measured. **That is [#139](https://github.com/danvisai/SDFusion/issues/139).**

### Pinned

`scripts/foundations/test_train_height_map_generator.py` — 167 tests, 17 of them this ticket's: `TestTypePrior` (3), `TestTypeStats` (5), `TestTypeTemperatureIsPreRegistered` (1), `TestLogitAdjustedTypeLoss` (4).

**Comment by danvisai** [comment]
Closed: NOT MET on `extra` and the collapse GUARD, KILL clause cleared for the first time. Answer recorded above and in `138-type-head-imbalance.md`; gist appended to map #1.
